server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -521,8 +535,20 @@ private synchronized ReleasableBytesReference pollOpsToWrite() {
     private void writeAndReleaseOps(ReleasableBytesReference toWrite) throws IOException {
         try (ReleasableBytesReference toClose = toWrite) {
             assert writeLock.isHeldByCurrentThread();
-            ByteBuffer ioBuffer = DiskIoBufferPool.getIoBuffer();
-
+            final int length = toWrite.length();
+            if (length == 0) {
+                return;
+            }
+            ByteBuffer ioBuffer = diskIoBufferPool.maybeGetDirectIOBuffer();
+            if (ioBuffer == null) {
+                // not using a direct buffer for writes from the current thread so just write without copying to the io buffer
+                BytesRefIterator iterator = toWrite.iterator();
+                BytesRef current;
+                while ((current = iterator.next()) != null) {
+                    Channels.writeToChannel(ByteBuffer.wrap(current.bytes, current.offset, current.length), channel);
```

Comment:
Could this not use the `byte[]` version of `witeToChannel` and thereby avoid one more byte-buffer allocation? That seems also shorter and easier to read.

```suggestion
                    Channels.writeToChannel(current.bytes, current.offset, current.length, channel);
```